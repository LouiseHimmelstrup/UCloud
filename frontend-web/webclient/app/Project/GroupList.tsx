import * as React from "react";
import * as Pagination from "@/Pagination";
import {Button, List, Icon, Flex} from "@/ui-components";
import * as Heading from "@/ui-components/Heading";
import {useHistory, useParams} from "react-router";
import GroupView from "./GroupView";
import {snackbarStore} from "@/Snackbar/SnackbarStore";
import {errorMessageOrDefault} from "@/UtilityFunctions";
import {usePromiseKeeper} from "@/PromiseKeeper";
import {Client} from "@/Authentication/HttpClientInstance";
import {NamingField} from "@/UtilityComponents";
import {ListRow} from "@/ui-components/List";
import {BreadCrumbsBase} from "@/ui-components/Breadcrumbs";
import {useRef, useState} from "react";
import {useCloudCommand} from "@/Authentication/DataHook";
import {Spacer} from "@/ui-components/Spacer";
import {Operation, Operations} from "@/ui-components/Operation";
import ProjectAPI, {isAdminOrPI, ProjectGroup, useGroupIdAndMemberId, useProjectFromParams} from "@/Project/Api";
import {bulkRequestOf} from "@/DefaultObjects";

export interface GroupWithSummary {
    groupId: string;
    groupTitle: string;
    numberOfMembers: number;
    members: string[];
}

const baseContext = "/projects/groups";

// UNUSED
const GroupList: React.FunctionComponent = () => {
    const history = useHistory();
    const {project, reload} = useProjectFromParams();
    const [groupId, membersPage] = useGroupIdAndMemberId();

    const allowManagement = isAdminOrPI(project?.status.myRole);

    const [creatingGroup, setCreatingGroup] = useState(false);
    const [, setLoading] = useState(false);
    const createGroupRef = useRef<HTMLInputElement>(null);
    const promises = usePromiseKeeper();
    const [renamingGroup, setRenamingGroup] = useState<string | null>(null);
    const renameRef = useRef<HTMLInputElement>(null);
    const [, runCommand] = useCloudCommand();

    const operations: GroupOperation[] = [
        {
            enabled: groups => groups.length === 1 && allowManagement,
            onClick: ([group]) => setRenamingGroup(group.id),
            icon: "rename",
            text: "Rename"
        },
        {
            enabled: groups => groups.length > 0 && allowManagement,
            onClick: (groups) => ProjectAPI.deleteGroup(bulkRequestOf(...groups.map(it => ({id: it.id})))),
            icon: "trash",
            text: "Delete",
            color: "red",
            confirm: true
        }
    ];

    const groups = project?.status.groups ?? [];

    if (groupId) return <GroupView />;
    const content = (
        <>
            {groups.length !== 0 || creatingGroup ? null : (
                <Flex justifyContent={"center"} alignItems={"center"} minHeight={"300px"} flexDirection={"column"}>
                    <Heading.h4>You have no groups to manage.</Heading.h4>
                    <ul>
                        <li>Groups are used to manage permissions in your project</li>
                        <li>You must create a group to grant members access to the project&#039;s files</li>
                    </ul>
                </Flex>
            )}
            <List>
                {creatingGroup ?
                    <ListRow
                        left={
                            <NamingField
                                confirmText="Create"
                                onSubmit={createGroup}
                                onCancel={() => setCreatingGroup(false)}
                                inputRef={createGroupRef}
                            />
                        }
                        leftSub={
                            <div />
                        }
                        right={<div />}
                        isSelected={false}
                        select={() => undefined}
                    /> : null}
                {groups.map((g, index) => (<React.Fragment key={g.id + index}>
                    <ListRow
                        left={
                            renamingGroup !== g.id ? g.specification.title : (
                                <NamingField
                                    confirmText="Rename"
                                    defaultValue={g.specification.title}
                                    onCancel={() => setRenamingGroup(null)}
                                    onSubmit={renameGroup}
                                    inputRef={renameRef}
                                />
                            )
                        }
                        navigate={() => {
                            if (renamingGroup !== g.id) {
                                history.push(`/project/members/${encodeURIComponent(g.id)}/${membersPage ?? ""}`);
                            }
                        }}
                        leftSub={<div />}
                        right={
                            <>
                                {g.status.members?.length === 0 ? null :
                                    <Flex>
                                        <Icon mt="4px" mr="4px" size="18" name="user" /> {g.status.members?.length}
                                    </Flex>
                                }

                                <Operations
                                    location={"IN_ROW"}
                                    operations={operations}
                                    entityNameSingular="Group"
                                    entityNamePlural="Groups"
                                    extra={{}}
                                    selected={[g]}
                                    row={g}
                                />
                            </>
                        }
                        isSelected={false}
                    />
                </React.Fragment>))}
            </List>
        </>
    );
    return <>
        <Spacer
            mb="12px"
            left={
                <BreadCrumbsBase embedded={false}>
                    <span>Groups</span>
                </BreadCrumbsBase>
            }

            right={!allowManagement ? null : (
                <Button mt="2px" height="40px" width="120px" onClick={() => setCreatingGroup(true)}>New Group</Button>
            )}
        />

        {content}
    </>;

    async function createGroup(e: React.SyntheticEvent): Promise<void> {
        e.preventDefault();
        try {
            setLoading(true);
            const groupName = createGroupRef.current?.value ?? "";
            if (!groupName) {
                snackbarStore.addFailure("Groupname can't be empty", false);
                return;
            }
            await promises.makeCancelable(Client.put(baseContext, {group: groupName})).promise;
            snackbarStore.addSuccess(`Group created`, true);
            createGroupRef.current!.value = "";
            setCreatingGroup(false);
            reload();
        } catch (err) {
            snackbarStore.addFailure(errorMessageOrDefault(err, "Could not create group."), false);
        } finally {
            setLoading(false);
        }
    }

    async function renameGroup(): Promise<void> {
        const groupId = renamingGroup;
        if (!groupId) return;
        const newGroupName = renameRef.current?.value;
        if (!newGroupName) return;

        const success = await runCommand(ProjectAPI.renameGroup(bulkRequestOf({group: groupId, newTitle: newGroupName})));

        if (!success) {
            snackbarStore.addFailure("Failed to rename project group", true);
            return;
        }

        reload();
        setRenamingGroup(null);
        snackbarStore.addSuccess("Project group renamed", true);
    }
};

type GroupOperation = Operation<ProjectGroup>;