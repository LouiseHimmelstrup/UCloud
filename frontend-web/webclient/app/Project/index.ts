import {APICallParameters} from "Authentication/DataHook";
import {buildQueryString} from "Utilities/URIUtilities";
import {Page, PaginationRequest} from "Types";
import {Client} from "Authentication/HttpClientInstance";
import {DEV_SITE, STAGING_SITE} from "../../site.config.json";
import {inDevEnvironment} from "UtilityFunctions";
import {IconName} from "ui-components/Icon";

const groupContext = "/projects/groups/";
const projectContext = "/projects/";

interface CreateGroupRequest {
    group: string;
}

export interface ListGroupMembersRequestProps extends PaginationRequest {
    group: string;
}

export function createGroup(props: CreateGroupRequest): APICallParameters<{}> {
    return {
        reloadId: Math.random(),
        method: "PUT",
        path: groupContext,
        parameters: props,
    };
}

export function listGroupMembersRequest(
    props: ListGroupMembersRequestProps
): APICallParameters<ListGroupMembersRequestProps> {
    return {
        method: "GET",
        path: buildQueryString(`${groupContext}members`, props),
        reloadId: Math.random(),
        parameters: props
    };
}

interface AddGroupMemberProps {
    group: string;
    memberUsername: string;
}

export function addGroupMember(payload: AddGroupMemberProps): APICallParameters<AddGroupMemberProps> {
    return {
        method: "PUT",
        path: `${groupContext}/members`,
        reloadId: Math.random(),
        payload
    };
}

interface RemoveGroupMemberProps {
    group: string;
    memberUsername: string;
}

export function removeGroupMemberRequest(payload: RemoveGroupMemberProps): APICallParameters<RemoveGroupMemberProps> {
    return {
        method: "DELETE",
        path: `${groupContext}members`,
        reloadId: Math.random(),
        payload
    };
}

export interface ShouldVerifyMembershipResponse {
    shouldVerify: boolean;
}

export function shouldVerifyMembership(projectId: string): APICallParameters {
    return {
        method: "GET",
        path: `${projectContext}should-verify`,
        reloadId: Math.random(),
        projectOverride: projectId
    };
}

export function verifyMembership(projectId: string): APICallParameters {
    return {
        method: "POST",
        path: `${projectContext}verify-membership`,
        reloadId: Math.random(),
        projectOverride: projectId
    };
}

export function projectRoleToString(role: ProjectRole): string {
    switch (role) {
        case ProjectRole.PI: return "PI";
        case ProjectRole.ADMIN: return "Admin";
        case ProjectRole.USER: return "User";
    }
}

export function projectStringToRole(role: string): ProjectRole {
    switch (role) {
        case "PI": return ProjectRole.PI;
        case "Admin": return ProjectRole.ADMIN;
        case "User": return ProjectRole.USER;
    }
    return ProjectRole.USER;
}

export function projectRoleToStringIcon(role: ProjectRole): IconName {
    switch (role) {
        case ProjectRole.PI: return "userPi";
        case ProjectRole.ADMIN: return "userAdmin";
        case ProjectRole.USER: return "user";
    }
}

export function groupSummaryRequest(payload: PaginationRequest): APICallParameters<PaginationRequest> {
    return {
        path: buildQueryString(`${groupContext}/summary`, payload),
        method: "GET",
        reloadId: Math.random(),
        parameters: payload,
        payload
    };
}

export interface UserStatusRequest {

}

export interface UserStatusResponse {
    membership: UserInProject[];
    groups: UserGroupSummary[];
}

export function userProjectStatus(request: UserStatusRequest): APICallParameters<UserStatusRequest> {
    return {
        reloadId: Math.random(),
        path: "/projects/membership",
        method: "POST",
        parameters: request,
        payload: request
    };
}

export interface MembershipSearchRequest extends PaginationRequest {
    query: string;
    notInGroup?: string;
}

type MembershipResponse = Page<ProjectMember>;

export function membershipSearch(request: MembershipSearchRequest): APICallParameters<MembershipSearchRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/membership/search", request),
        parameters: request
    };
}

export interface UpdateGroupNameRequest {
    oldGroupName: string;
    newGroupName: string;
}

export function updateGroupName(request: UpdateGroupNameRequest): APICallParameters<UpdateGroupNameRequest> {
    return {
        method: "POST",
        path: "/projects/groups/update-name",
        parameters: request,
        payload: request
    };
}

export interface DeleteGroupRequest {
    groups: string[];
}

export function deleteGroup(request: DeleteGroupRequest): APICallParameters<DeleteGroupRequest> {
    return {
        method: "DELETE",
        path: "/projects/groups",
        parameters: request,
        payload: request
    };
}

export interface ProjectMember {
    username: string;
    role: ProjectRole;
}

export interface Project {
    id: string;
    title: string;
    members: ProjectMember[];
}

export const emptyProject = (id: string): Project => ({
    id,
    title: "",
    members: []
});

export enum ProjectRole {
    PI = "PI",
    ADMIN = "ADMIN",
    USER = "USER"
}

export interface UserInProject {
    projectId: string;
    title: string;
    whoami: ProjectMember;
    needsVerification: boolean;
    favorite: boolean;
    archived: boolean;
}

export interface UserGroupSummary {
    projectId: string;
    group: string;
    username: string;
}

export const createProject = (payload: {title: string}): APICallParameters => ({
    method: "POST",
    path: "/projects",
    payload,
    reloadId: Math.random()
});

export const inviteMember = (payload: {projectId: string; usernames: string[]}): APICallParameters => ({
    method: "POST",
    path: "/projects/invites",
    payload,
    reloadId: Math.random()
});

export const deleteMemberInProject = (payload: {projectId: string; member: string}): APICallParameters => ({
    method: "DELETE",
    path: "/projects/members",
    payload,
    reloadId: Math.random()
});

export const changeRoleInProject = (
    payload: {projectId: string; member: string; newRole: ProjectRole}
): APICallParameters => ({
    method: "POST",
    path: "/projects/members/change-role",
    payload,
    reloadId: Math.random()
});

export interface ListProjectsRequest extends PaginationRequest{
    archived?: boolean;
    noFavorites?: boolean;
}

export const listProjects = (parameters: ListProjectsRequest): APICallParameters<ListProjectsRequest> => ({
    method: "GET",
    path: buildQueryString(
        "/projects/list",
        parameters
    ),
    parameters,
    reloadId: Math.random()
});

export interface ListFavoriteProjectsRequest extends PaginationRequest {
    archived: boolean;
}

export const listFavoriteProjects = (parameters: ListFavoriteProjectsRequest): APICallParameters<ListFavoriteProjectsRequest> => ({
    method: "GET",
    path: buildQueryString(
        "/projects/listFavorites",
        parameters
    ),
    parameters,
    reloadId: Math.random()
});

export const roleInProject = (project: ProjectMember[]): ProjectRole | undefined => {
    const member = project.find(m => {
        return m.username === Client.username;
    });

    if (member === undefined) return undefined;
    return member.role;
};


export interface ToggleProjectFavorite{
    projectId: string;
}

export function toggleFavoriteProject(request: ToggleProjectFavorite): APICallParameters<ToggleProjectFavorite> {
    return {
        method: "POST",
        path: "/projects/favorite",
        parameters: request,
        payload: request
    };
}

export interface OutgoingInvite {
    username: string;
    invitedBy: string;
    timestamp: number;
}

export interface ListOutgoingInvitesRequest extends PaginationRequest {

}

export function listOutgoingInvites(request: ListOutgoingInvitesRequest): APICallParameters<ListOutgoingInvitesRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/invites/outgoing", request),
        parameters: request
    };
}

export interface IngoingInvite {
    project: string;
    invitedBy: string;
    timestamp: string;
}

export interface ListIngoingInvitesRequest extends PaginationRequest {

}

export function listIngoingInvites(request: ListIngoingInvitesRequest): APICallParameters<ListIngoingInvitesRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/invites/ingoing", request),
        parameters: request
    };
}

export interface AcceptInviteRequest {
    projectId: string;
}

export function acceptInvite(request: AcceptInviteRequest): APICallParameters<AcceptInviteRequest> {
    return {
        method: "POST",
        path: "/projects/invites/accept",
        payload: request,
        parameters: request
    };
}

export interface RejectInviteRequest {
    projectId: string;
    username?: string;
}

export function rejectInvite(request: RejectInviteRequest): APICallParameters<RejectInviteRequest> {
    return {
        method: "DELETE",
        path: "/projects/invites/reject",
        parameters: request,
        payload: request
    };
}

type LeaveProjectRequest = {};
export function leaveProject(request: LeaveProjectRequest): APICallParameters<LeaveProjectRequest> {
    return {
        method: "DELETE",
        path: "/projects/leave",
        parameters: request,
        payload: request
    };
}

export interface TransferPiRoleRequest {
    newPrincipalInvestigator: string;
}

export function transferPiRole(request: TransferPiRoleRequest): APICallParameters<TransferPiRoleRequest> {
    return {
        method: "POST",
        path: "/projects/transfer-pi",
        parameters: request,
        payload: request
    };
}

export type ListRepositoryRequest = PaginationRequest;

export function listRepositoryFiles(
    request: ListRepositoryRequest,
    projectOverride?: string
): APICallParameters<ListRepositoryRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/repositories/list-files", request),
        parameters: request,
        reloadId: Math.random(),
        projectOverride
    };
}

export interface ArchiveProjectRequest {
    archiveStatus: boolean;
}

export function setProjectArchiveStatus(request: ArchiveProjectRequest): APICallParameters<ArchiveProjectRequest> {
    return {
        method: "POST",
        path: "/projects/archive",
        parameters: request,
        payload: request,
        reloadId: Math.random()
    };
}

export interface ViewProjectRequest {
    id: string;
}

export function viewProject(request: ViewProjectRequest): APICallParameters<ViewProjectRequest> {
    return {
        method: "GET",
        path: buildQueryString("/projects/view", request),
        parameters: request,
        reloadId: Math.random()
    };
}

export function areProjectsEnabled(): boolean {
    if ([DEV_SITE, STAGING_SITE].includes(window.location.host) || inDevEnvironment()) {
        return true;
    }
    return Client.userRole === "ADMIN";
}
