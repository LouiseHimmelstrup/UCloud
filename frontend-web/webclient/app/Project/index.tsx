import { Contributor, Subject, RelatedIdentifier } from "./api";
import { RouterLocationProps } from "Utilities/URIUtilities";
import { AddSnackOperation } from "Snackbar/Snackbars";
export { default as CreateUpdate } from "./CreateUpdate";
export { ManagedView, View } from "./View";
export { default as Manage } from "./Management"

export interface CreateUpdateState {
    id?: string
    path: string
    title: string
    description: string
    license?: { title: string, link: string, identifier: string }
    keywords: string[]
    notes: string
    dataManagementPlan: string
    contributors: Contributor[]
    references: string[]
    grants: string[]
    subjects: Subject[]
    relatedIdentifiers: RelatedIdentifier[]
    errors: {
        description?: boolean
        title?: boolean
        contributors: {}
        subjects: {}
        relatedIdentifiers: {}
    }
}

export type CreateUpdateProps = RouterLocationProps & AddSnackOperation;

export interface DetailedProjectSearchReduxState {
    error?: string
    loading: boolean
    hidden: boolean
    projectName: string
}

export interface DetailedProjectSearchOperations {
    setProjectName: (name: string) => void
    fetchProjectsFromName: (name: string, itemsPerPage: number, page: number, callback: () => void) => void
    setError: (error?: string) => void
}