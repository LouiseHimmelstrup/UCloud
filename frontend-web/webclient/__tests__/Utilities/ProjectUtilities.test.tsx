import * as ProjectUtils from "../../app/Utilities/ProjectUtilities";
import {ProjectRole} from "../../app/Project";
import {GroupWithSummary} from "../../app/Project/GroupList";
let groupCounter = 0;

function newGroupWithSummary(): GroupWithSummary {
    return {
        groupId: groupCounter.toString(),
        groupTitle: "Group " + groupCounter++,
        members: ["foo", "bar", "baz"],
        numberOfMembers: 3
    };
}

jest.mock("Authentication/DataHook", () => ({
    useCloudAPI: (path) => ([{data: groupSummaryPage, loading: false, error: undefined}])
}));

const groupSummaryPage: Page<GroupWithSummary> = {
    items: [newGroupWithSummary(), newGroupWithSummary(), newGroupWithSummary()],
    itemsInTotal: 3,
    itemsPerPage: 25,
    pageNumber: 0
};

describe("isAdminOrPI", () => {
    test("User", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.USER)).toBe(false);
    });

    test("Admin", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.ADMIN)).toBe(true);
    });

    test("PI", () => {
        expect(ProjectUtils.isAdminOrPI(ProjectRole.PI)).toBe(true);
    });
});
