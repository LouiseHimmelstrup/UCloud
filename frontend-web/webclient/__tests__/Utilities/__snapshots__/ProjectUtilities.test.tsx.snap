// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdatePermissionsDialog Mount 1`] = `
.c0 {
  width: auto;
  min-width: 450px;
}

.c15 {
  margin-left: auto;
}

.c14 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--red,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c14:disabled {
  opacity: 0.25;
}

.c14:focus {
  outline: none;
}

.c14:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c16 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--blue,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c16:disabled {
  opacity: 0.25;
}

.c16:focus {
  outline: none;
}

.c16:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c10 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 512px;
  margin-bottom: 512px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 25px;
}

.c3 {
  cursor: inherit;
  width: 300px;
  margin-right: 16px;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-template-columns: repeat(auto);
  -webkit-column-gap: 5px;
  column-gap: 5px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  border-radius: 5px;
  border: 1px solid var(--invertedThemeColor);
  color: var(--invertedThemeColor,#f00);
  -webkit-transition: all 300ms ease;
  transition: all 300ms ease;
}

.c6 {
  position: relative;
  height: 40px;
  width: 40px;
  -webkit-transition: -webkit-transform 300ms ease;
  -webkit-transition: transform 300ms ease;
  transition: transform 300ms ease;
}

.c12 {
  position: relative;
  height: 40px;
  width: 40px;
  -webkit-transition: -webkit-transform 300ms ease;
  -webkit-transition: transform 300ms ease;
  transition: transform 300ms ease;
}

.c12:hover {
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c12:hover:hover > .c8 {
  color: var(--blue,#f00);
  border: 1px solid var(--blue,#f00);
}

.c7 {
  opacity: 0;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  margin: 0;
  cursor: pointer;
}

.c7:checked + .c8 {
  background-color: var(--blue,#f00);
  border: 0px solid white;
  color: white;
}

.c11 {
  text-align: center;
  font-size: 0.5em;
  font-weight: 600;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1;
  padding-top: 0.1rem;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c1 {
  height: 300px;
  overflow-y: auto;
}

<div
  className="c0"
  width="auto"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3 c4"
        cursor="inherit"
        title="Group 0"
        width="300px"
      >
        Group 0
      </div>
      <div
        className="c5"
      >
        <div
          checked={true}
          className="c6"
          height={40}
        >
          <input
            checked={true}
            className="c7"
            id="None"
            name="0"
            onChange={[Function]}
            type="radio"
            value="None"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 25 24"
              width="65%"
            >
              <path
                d="M24.002 2.001L22 .001 12 9.6l-10-9.6-2 2 9.6 10L.001 22l2 2 10-9.6 10 9.6 2-2-9.6-10 9.6-9.999z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="None"
            >
              None
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Read"
            name="0"
            onChange={[Function]}
            type="radio"
            value="Read"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              cursor="inherit"
              fill="currentcolor"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M15.5 14h-.8l-.3-.3c1-1.1 1.6-2.6 1.6-4.2C16 5.9 13.1 3 9.5 3S3 5.9 3 9.5 5.9 16 9.5 16c1.6 0 3.1-.6 4.2-1.6l.3.3v.8l5 5 1.5-1.5-5-5zm-6 0C7 14 5 12 5 9.5S7 5 9.5 5 14 7 14 9.5 12 14 9.5 14z"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Read"
            >
              Read
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Edit"
            name="0"
            onChange={[Function]}
            type="radio"
            value="Edit"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M23.6 5.4a1.289 1.289 0 000-1.867L20.467.4A1.289 1.289 0 0018.6.4l-2.467 2.467 5 5L23.6 5.4z"
                fill="currentcolor"
                fillRule="nonzero"
              />
              <path
                d="M0 19v5h5L19.733 9.267l-5-5L0 19z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Edit"
            >
              Edit
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3 c4"
        cursor="inherit"
        title="Group 1"
        width="300px"
      >
        Group 1
      </div>
      <div
        className="c5"
      >
        <div
          checked={true}
          className="c6"
          height={40}
        >
          <input
            checked={true}
            className="c7"
            id="None"
            name="1"
            onChange={[Function]}
            type="radio"
            value="None"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 25 24"
              width="65%"
            >
              <path
                d="M24.002 2.001L22 .001 12 9.6l-10-9.6-2 2 9.6 10L.001 22l2 2 10-9.6 10 9.6 2-2-9.6-10 9.6-9.999z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="None"
            >
              None
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Read"
            name="1"
            onChange={[Function]}
            type="radio"
            value="Read"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              cursor="inherit"
              fill="currentcolor"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M15.5 14h-.8l-.3-.3c1-1.1 1.6-2.6 1.6-4.2C16 5.9 13.1 3 9.5 3S3 5.9 3 9.5 5.9 16 9.5 16c1.6 0 3.1-.6 4.2-1.6l.3.3v.8l5 5 1.5-1.5-5-5zm-6 0C7 14 5 12 5 9.5S7 5 9.5 5 14 7 14 9.5 12 14 9.5 14z"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Read"
            >
              Read
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Edit"
            name="1"
            onChange={[Function]}
            type="radio"
            value="Edit"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M23.6 5.4a1.289 1.289 0 000-1.867L20.467.4A1.289 1.289 0 0018.6.4l-2.467 2.467 5 5L23.6 5.4z"
                fill="currentcolor"
                fillRule="nonzero"
              />
              <path
                d="M0 19v5h5L19.733 9.267l-5-5L0 19z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Edit"
            >
              Edit
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      className="c2"
    >
      <div
        className="c3 c4"
        cursor="inherit"
        title="Group 2"
        width="300px"
      >
        Group 2
      </div>
      <div
        className="c5"
      >
        <div
          checked={true}
          className="c6"
          height={40}
        >
          <input
            checked={true}
            className="c7"
            id="None"
            name="2"
            onChange={[Function]}
            type="radio"
            value="None"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 25 24"
              width="65%"
            >
              <path
                d="M24.002 2.001L22 .001 12 9.6l-10-9.6-2 2 9.6 10L.001 22l2 2 10-9.6 10 9.6 2-2-9.6-10 9.6-9.999z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="None"
            >
              None
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Read"
            name="2"
            onChange={[Function]}
            type="radio"
            value="Read"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              cursor="inherit"
              fill="currentcolor"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M15.5 14h-.8l-.3-.3c1-1.1 1.6-2.6 1.6-4.2C16 5.9 13.1 3 9.5 3S3 5.9 3 9.5 5.9 16 9.5 16c1.6 0 3.1-.6 4.2-1.6l.3.3v.8l5 5 1.5-1.5-5-5zm-6 0C7 14 5 12 5 9.5S7 5 9.5 5 14 7 14 9.5 12 14 9.5 14z"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Read"
            >
              Read
            </label>
          </div>
        </div>
        <div
          checked={false}
          className="c12"
          height={40}
        >
          <input
            checked={false}
            className="c7"
            id="Edit"
            name="2"
            onChange={[Function]}
            type="radio"
            value="Edit"
          />
          <div
            className="c8 c9"
          >
            <svg
              className="c10"
              clipRule="evenodd"
              cursor="inherit"
              fill="currentcolor"
              fillRule="evenodd"
              height="65%"
              viewBox="0 0 24 24"
              width="65%"
            >
              <path
                d="M23.6 5.4a1.289 1.289 0 000-1.867L20.467.4A1.289 1.289 0 0018.6.4l-2.467 2.467 5 5L23.6 5.4z"
                fill="currentcolor"
                fillRule="nonzero"
              />
              <path
                d="M0 19v5h5L19.733 9.267l-5-5L0 19z"
                fillRule="nonzero"
              />
            </svg>
            <label
              className="c11"
              fontSize="0.5em"
              htmlFor="Edit"
            >
              Edit
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c13"
  >
    <button
      className="c14"
      color="red"
      onClick={[Function]}
    >
      Cancel
    </button>
    <div
      className="c15"
    />
    <button
      className="c16"
      color="blue"
      disabled={true}
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
